src/microbe_stage/systems/MicrobeProcessManagerSystem.cs

```diff
@@ -0,0 +1,180 @@
+    /// </remarks>
+    [With(typeof(CompoundStorage))]
+    [With(typeof(BioProcesses))]
+    public sealed class MicrobeProcessManagerSystem : AEntitySetSystem<float>
+    {
+        private static readonly Compound ATP = SimulationParameters.Instance.GetCompound("atp");
+        private static readonly Compound Glucose = SimulationParameters.Instance.GetCompound("glucose");
+        private Dictionary<BioProcess, float> cachedPriority = new();
+        private BiomeConditions? biome;
+
+        public MicrobeProcessManagerSystem(World world, IParallelRunner runner) : base(world, runner)
+        {
+        }
+
+        /// <summary>
+        ///   Sets the biome whose environmental values affect processes
+        /// </summary>
+        public void SetBiome(BiomeConditions newBiome)
+        {
+            biome = newBiome;
+        }
+
+        protected override void Update(float delta, in Entity entity)
+        {
+            ref var storage = ref entity.Get<CompoundStorage>();
+            ref var processes = ref entity.Get<BioProcesses>();
+
+            ManageProcesses(ref processes, ref storage, delta);
+        }
+
+        private float GetPriority(BioProcess process)
+        {
+            var key = process;
+
+            if (cachedPriority.TryGetValue(key, out var cached))
+            {
+                return cached;
+            }
+
+            // Lower priorities get to run first, so start at bottom
+            // It is important non-ATP-producers run first
+            // Different levels of priority "coarseness" are stored in different place values
+            var priority = 0.0f;
+
+            var inputAmount = 0.0f;
+            foreach (var entry in process.Inputs)
+            {
+                if (entry.Key.IsEnvironmental)
+                    continue;
+
+                inputAmount += entry.Value;
+            }
+
+            if (process.Outputs.TryGetValue(ATP, out var atpOut))
+            {
+                if (inputAmount <= 0)
+                {
+                    priority = 0;
+                }
+                else
+                {
+                    priority = 10 - atpOut / inputAmount / 100;
+                }
+            }
+
+            // Penalize Glucose consumers to allow RuBisCo to work
+            if (process.Inputs.ContainsKey(Glucose))
+            {
+                priority += 100;
+            }
+
+            cachedPriority.Add(key, priority);
+            return priority;
+        }
+
+        private void ManageProcesses(ref BioProcesses processor, ref CompoundStorage storage, float delta)
+        {
+            // Make sure there are ActiveProcesses
+            if (processor.ActiveProcesses == null)
+                return;
+
+            if (biome == null)
+                throw new NullReferenceException("Biome needs to be set");
+
+            var glucoseConsuming = false;
+
+            // Create dummy compound bag
+            var bag = storage.Compounds.Clone();
```

Comment:
This is a no-go, allocating duplicate compounds bag for each entity on each game update. This is going to cause so much GC garbage.

Reply:
The exact method of how I implemented the RuBisCo logic here changed several times. I know how to remove this extra compound bag, but it's at the cost of future ease of editing. I overall agree that it would be a problem as is.

Reply:
It needs to be done, unless you are fine with the process speed control running only like once per second for each microbe... With the ECS refactor I've been way more careful than before to ensure that no operation that runs each frame allocates memory (and the few places where that couldn't be avoided use an array pool).